Source/WebCore/loader/DocumentPrefetcher.cpp

```diff
@@ -0,0 +1,229 @@
+        FetchOptions::Mode::Navigate,
+        CertificateInfoPolicy::IncludeCertificateInfo,
+        ContentSecurityPolicyImposition::DoPolicyCheck,
+        DefersLoadingPolicy::AllowDefersLoading,
+        CachingPolicy::AllowCachingPrefetch
+    );
+    prefetchOptions.destination = FetchOptions::Destination::Document;
+    CachedResourceRequest prefetchRequest(WTFMove(request), prefetchOptions);
+    if (lowPriority)
+        prefetchRequest.setPriority(ResourceLoadPriority::Low);
+
+    auto resourceErrorOr = document->protectedCachedResourceLoader()->requestRawResource(WTFMove(prefetchRequest));
+
+    if (!resourceErrorOr)
+        return;
+    auto prefetchedResource = resourceErrorOr.value();
+    if (prefetchedResource) {
+        m_prefetchedResources.set(url, prefetchedResource);
+        prefetchedResource->addClient(*this);
+    }
+}
+
+void DocumentPrefetcher::responseReceived(const CachedResource&, const ResourceResponse&, CompletionHandler<void()>&& completionHandler)
+{
+    if (completionHandler)
+        completionHandler();
+}
+
+void DocumentPrefetcher::redirectReceived(CachedResource& resource, ResourceRequest&& request, const ResourceResponse&, CompletionHandler<void(ResourceRequest&&)>&& completionHandler)
+{
+    URL originalURL;
+    for (auto& [url, prefetchedResource] : m_prefetchedResources) {
+        if (prefetchedResource.get() == &resource) {
+            originalURL = url;
+            break;
+        }
+    }
+
+    if (!originalURL.isEmpty()) {
+        URL redirectURL = request.url();
+
+        if (m_notifyWhenFinishedURLs.contains(originalURL)) {
+            if (auto prefetchedResource = m_prefetchedResources.take(originalURL))
+                m_prefetchedResources.set(redirectURL, prefetchedResource);
+
+            if (auto metrics = m_prefetchedNetworkLoadMetrics.take(originalURL))
+                m_prefetchedNetworkLoadMetrics.set(redirectURL, WTFMove(metrics));
+        }
+
+        bool hadPendingNotification = m_notifyWhenFinishedURLs.remove(originalURL);
+        if (hadPendingNotification)
+            m_notifyWhenFinishedURLs.add(redirectURL);
+    }
+
+    completionHandler(WTFMove(request));
+}
+
+void DocumentPrefetcher::notifyFinished(CachedResource& resource, const NetworkLoadMetrics& metrics, LoadWillContinueInAnotherProcess)
+{
+    m_finishedURLs.add(resource.url());
+    URL completedURL;
+    for (auto& [url, prefetchedResource] : m_prefetchedResources) {
+        if (prefetchedResource.get() == &resource && !m_prefetchedNetworkLoadMetrics.contains(url)) {
+            m_prefetchedNetworkLoadMetrics.set(url, Box<NetworkLoadMetrics>::create(metrics));
+            completedURL = url;
+            break;
+        }
+    }
+
+    if (resource.hasClient(*this))
+        resource.removeClient(*this);
+}
+
+void DocumentPrefetcher::clearPrefetchedAssets()
+{
+#if ASSERT_ENABLED
+    for (auto& [url, prefetchedResource] : m_prefetchedResources) {
```

Comment:
nit: this could just iterate m_prefetchedResources.values() since the key is unused.